core/storage/topic/gae_models_test.py

```diff
@@ -0,0 +1,147 @@
+# coding: utf-8
+#
+# Copyright 2018 The Oppia Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Test for Topic model."""
+
+from constants import constants
+from core.domain import subtopic_page_domain
+from core.domain import subtopic_page_services
+from core.domain import topic_domain
+from core.domain import topic_services
+from core.platform import models
+from core.tests import test_utils
+import feconf
+
+(topic_models,) = models.Registry.import_models([models.NAMES.topic])
+
+
+class TopicModelUnitTest(test_utils.GenericTestBase):
+    """Test the TopicModel class."""
+
+    def setUp(self):
+        super(TopicModelUnitTest, self).setUp()
+        self.topic = topic_domain.Topic.create_default_topic(
+            topic_id='topic_id',
+            name='topic_test'
+        )
+        topic_services.save_new_topic(feconf.SYSTEM_COMMITTER_ID, self.topic)
+
+    def test__trusted_commit(self):
```

Comment:
Can you explain what exactly this is testing?

For example, test_get_by_name can be explained as: "Test that topics can be correctly retrieved by name." There's a clear way for that to fail, e.g. when trying to retrieve a topic by name, the wrong topic is retrieved or no topic is retrieved. I'm not sure what the corresponding equivalent for this one is. Note also that, in general, private methods don't need their own test method (cf other comment).

Also this sort of thinking would suggest other tests that are missing. E.g. we should also test the behaviour that arises when there is no topic with the given name. Tests should be comprehensive; coverage is just a simple tool for checking that all lines of code in the module being tested have run at least once. So, try to figure out what behaviours are actually being exhibited so that you can write tests that ensure that those behaviours remain correct even through subsequent code changes.

It might also not be a bad idea to write a one-line docstring for each test method, to ensure there's clarity on what is being tested.

Reply:
Ok, when running my coverage report it said that there were missing lines to be tested inside _trusted_commit() method. So, I didn't know what to test as that method doesn't return anything, I asked on gitter and they answered me that I should test the variables that are assigned on that method and check that their value is correct. After looking at the code I realized that the only variable that I can access is version as it's a class variable and I could access it from by class instance. I wasn't convinced about testing that method however if I don't test it then coverage report doesn't give me 100% coverage

Reply:
_trusted_commit() is a private method, similar to _get_instance_id() which we talked about earlier. In the discussion for that private method, I explained how to find a public method that uses that private method, and write tests against that instead, ensuring that the tests exercise the code paths in the private method. Could you apply a similar process of reasoning here?

Reply:
Sorry! My previous comment was misleading. For some reason I was thinking about the previous issue with private methods when I wrote it, my bad.

I took a look at the code to see what exactly that method is doing. You're right that it increments the version and I think it's a good idea to check that. However, the more important thing that it does is that it actually saves stuff to the datastore (i.e. it has side-effects). 

So, what I would suggest is that you do the following: after calling the commit method, try to retrieve the newly-added models from the datastore and verify that something's retrieved in both cases. Then, before calling the commit method, add similar calls to retrieve those models from the datastore and verify that there's nothing there beforehand, so that you can be certain that the commit() actually caused the models to be written to the datastore.

Does that sound good?

Reply:
Also you might want to rename the test to something that more clearly expresses what it's doing, e.g. test_that_subsidiary_models_are_created_when_new_model_is_saved (or similar).

Reply:
Something like this?
``` python
    def test_that_subsidiary_models_are_created_when_new_model_is_saved(self):
        """ Test the _trusted_commit() method """

        # Topic is created but not commited/saved
        topic = topic_models.TopicModel(
            id='topic_id',
            name='topic_name',
            subtopic_schema_version=feconf.CURRENT_SUBTOPIC_SCHEMA_VERSION,
            next_subtopic_id=1,
            language_code='en'
        )
        # We check that topic has not been saved before calling commit()
        self.assertIsNone(topic_models.TopicModel.get_by_name('topic_name'))
        # We call commit() expecting that _trusted_commit works fine
        # and saves topic to datastore
        topic.commit(
            committer_id=feconf.SYSTEM_COMMITTER_ID,
            commit_message='Created new topic',
            commit_cmds=[{'cmd': topic_domain.CMD_CREATE_NEW}]
        )
        # Now we check that topic is not None and that actually
        # now topic exists, that means that commit() worked fine
        self.assertIsNotNone(topic_models.TopicModel.get_by_name('topic_name'))
```

Reply:
I applied the test above in this commit: https://github.com/oppia/oppia/pull/5689/commits/64de9c8d0e165d4979f4ea006efba801158ba52b